drivers/ieee802154/ieee802154_gecko.c

```diff
@@ -0,0 +1,588 @@
+		.txToRxSearchTimeout = 0,
+	},
+	.framesMask = RAIL_IEEE802154_ACCEPT_STANDARD_FRAMES,
+	.promiscuousMode = false,
+	.isPanCoordinator = false,
+};
+
+static void ieee802154_gecko_rail_cb(RAIL_Handle_t rail_handle,
+				     RAIL_Events_t event);
+
+static RAIL_Config_t rail_config = {
+	.eventsCallback = &ieee802154_gecko_rail_cb,
+};
+
+static void ieee802154_gecko_get_eui64(u8_t *mac)
+{
+	u64_t uniqueID = SYSTEM_GetUnique();
+
+	memcpy(mac, &uniqueID, sizeof(uniqueID));
+}
+
+static void ieee802154_gecko_rx(struct device *dev,
+				RAIL_RxPacketHandle_t packet_handle,
+				RAIL_RxPacketInfo_t *packet_info)
+{
+	struct ieee802154_gecko_context *dev_data = DEV_DATA(dev);
+	struct net_pkt *pkt;
+
+	LOG_DBG("Rx packet received");
+
+	pkt = net_pkt_alloc_with_buffer(dev_data->iface,
+					packet_info->packetBytes,
+					AF_UNSPEC, 0, K_NO_WAIT);
+	if (pkt == NULL) {
+		LOG_ERR("No net_pkt available");
+		return;
+	}
+
+	/* Skip Frame Length field, 1 byte at index 0 */
+	if (net_pkt_write(pkt, packet_info->firstPortionData + 1,
+			  packet_info->firstPortionBytes - 1)) {
+		goto drop;
+	}
+	if (net_pkt_write(pkt, packet_info->lastPortionData,
+			  packet_info->packetBytes - packet_info->firstPortionBytes)) {
+		goto drop;
+	}
+
+	/* Fill packet information */
+	RAIL_RxPacketDetails_t packet_details = {
+		.timeReceived = {
+			.timePosition = RAIL_PACKET_TIME_AT_SYNC_END,
+		},
+	};
+	RAIL_GetRxPacketDetails(dev_data->rail_handle, packet_handle, &packet_details);
+
+	net_pkt_set_ieee802154_lqi(pkt, packet_details.lqi);
+	net_pkt_set_ieee802154_rssi(pkt, packet_details.rssi);
+
+#if defined(CONFIG_NET_PKT_TIMESTAMP)
+	struct net_ptp_time timestamp = {
+		.second = packet_details.timeReceived.packetTime / USEC_PER_SEC,
+		.nanosecond =
+			(packet_details.timeReceived.packetTime % USEC_PER_SEC)
+			* NSEC_PER_USEC
+	};
+
+	net_pkt_set_timestamp(pkt, &timestamp);
+#endif
+
+	if (net_recv_data(dev_data->iface, pkt) < 0) {
+		LOG_ERR("Packet dropped by NET stack");
+		goto drop;
+	}
+
+	return;
+
+drop:
+	net_pkt_unref(pkt);
+}
+
+/* Radio device API */
+
+static void ieee802154_gecko_iface_init(struct net_if *iface)
+{
+	struct device *dev = net_if_get_device(iface);
+	struct ieee802154_gecko_context *dev_data = DEV_DATA(dev);
+
+	ieee802154_gecko_get_eui64(dev_data->mac);
+	net_if_set_link_addr(iface, dev_data->mac, sizeof(dev_data->mac),
+			     NET_LINK_IEEE802154);
+
+	dev_data->iface = iface;
+
+	ieee802154_init(iface);
+}
+
+static enum ieee802154_hw_caps ieee802154_gecko_get_capabilities(struct device *dev)
+{
+	return IEEE802154_HW_FCS |
+	       IEEE802154_HW_FILTER |
+	       IEEE802154_HW_CSMA |
+	       IEEE802154_HW_2_4_GHZ;
+}
+
+static int ieee802154_gecko_cca(struct device *dev)
+{
+	return 0;
+}
+
+static int ieee802154_gecko_set_channel(struct device *dev, u16_t channel)
+{
+	struct ieee802154_gecko_context *dev_data = DEV_DATA(dev);
+	RAIL_Status_t status;
+
+	status = RAIL_PrepareChannel(dev_data->rail_handle, channel);
```

Comment:
From the RAIL API: "When successful, the radio is idled. When unsuccessful, the radio state will not be altered.". Hence, the radio will become idled, is this wanted? Otherwise, removing the RAIL_PrepareChannel should be okay I guess.